### PR TITLE
[diffusion] fix Z-Image SP sharding for portrait and padded resolutions

### PR DESCRIPTION
**Pull request by Ratish1**
## Motivation

   Fix Z-Image-Turbo sequence-parallel sharding for portrait and padded resolutions when Ulysses/SP is enabled. Fixes #21021

Z-Image currently works for some resolutions such as 1024x1024 and 1280x720, but produces corrupted results for others such as 720x1280 and 720x720 when sharding is enabled. This PR fixes the Z-Image-specific SP path so it preserves native image geometry during denoising.

Z-Image patchifies latents in native F/H/W order, but the current Z-Image SP path mutates the image geometry before
denoising:

- it swaps H/W so sharding always happens along the larger spatial axis
- it pads fake spatial rows so each rank gets the same local shard height

## Modifications

- Remove the Z-Image H/W swap path and keep sharding in native geometry.
- Build a per-request SP plan that chooses either native height or native width splitting based on lower token-padding cost.
- Shard rectangular latent slices directly along the chosen native axis.
- Gather uneven local shards explicitly with all_gather, then crop each rank’s contribution back to its true size before
  concatenation.
- Pass the Z-Image batch through the SP gather path so final gather can use the per-request shard plan.
- Pad only the local patch-token sequence in the Z-Image DiT path to the aligned target length required by Ulysses.
- Keep the shared gather_latents_for_sp(..., batch=None) signature compatible for other pipelines.

## Accuracy Tests

I tested with 2 commands:
`sglang generate --model-path Tongyi-MAI/Z-Image-Turbo --seed 42 --prompt "A crowded beach" --height 720 --width 720 --num-inference-steps 9 --ulysses-degree 2 --num-gpus 2 --guidance-scale 4.0`

`sglang generate --model-path Tongyi-MAI/Z-Image-Turbo --seed 42 --prompt "A crowded beach" --height 720 --width 1280 --num-inference-steps 9 --ulysses-degree 2 --num-gpus 2 --guidance-scale  4.0`

Before PR:
image

image

After PR:
image
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py [hunk 1]
```diff
@@ -813,7 +813,9 @@ def _postprocess_sp_latents(
     ) -> tuple[torch.Tensor, torch.Tensor | None]:
         """Gather latents after Sequence Parallelism if they were sharded."""
         if get_sp_world_size() > 1 and getattr(batch, "did_sp_shard_latents", False):
-            latents = self.server_args.pipeline_config.gather_latents_for_sp(latents)
+            latents = self.server_args.pipeline_config.gather_latents_for_sp(
+                latents, batch=batch
+            )
             if trajectory_tensor is not None:
                 # trajectory_tensor shapes:
                 # - video: [b, num_steps, c, t_local, h, w] -> gather on dim=3
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:818]: This still assumes SP latents are always gathered along a fixed dim. For Z-Image, this PR can shard along W as well, so `return_trajectory_latents` will reconstruct the wrong tensor on the W-shard path. Should this reuse the same gather logic as final latents?

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py [hunk 2]
```diff
@@ -693,7 +700,11 @@ def patchify_and_embed(
             F_tokens * H_tokens * W_tokens, pF * pH * pW * C
         )
         image_ori_len = image.size(0)
-        image_padding_len = (-image_ori_len) % SEQ_MULTI_OF
+        image_seq_len_target = max(
+            self._ceil_to_multiple(image_ori_len, SEQ_MULTI_OF),
+            image_seq_len_target or 0,
+        )
+        image_padding_len = image_seq_len_target - image_ori_len
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/zimage.py:776]: Can we assume `image_padding_len` is always >=0?

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py [hunk 3]
```diff
@@ -652,12 +652,19 @@ def create_coordinate_grid(size, start=None, device=None):
         grids = torch.meshgrid(axes, indexing="ij")
         return torch.stack(grids, dim=-1)
 
+    @staticmethod
+    def _ceil_to_multiple(value: int, multiple: int) -> int:
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/dits/zimage.py:722]: Should we remove this helper func to `utils.py`?

**Comment by BBuf** [comment]
This PR fixes the portrait / padded-resolution / W-shard path, but I don't see a regression test that actually exercises that path. The current 2-GPU Z-Image test still uses the default square case, so it likely never hits the new branch. Could we add one portrait or padded-resolution SP test here?

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py (hunk 1 above)
Comment (Ratish1) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/denoising.py:818]: Yes, correct. fixed it now

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py (hunk 3 above)
Comment (Ratish1) [reply python/sglang/multimodal_gen/runtime/models/dits/zimage.py:722]: Should we move it since its just a helper func specific to this file?

File: python/sglang/multimodal_gen/runtime/models/dits/zimage.py (hunk 2 above)
Comment (Ratish1) [reply python/sglang/multimodal_gen/runtime/models/dits/zimage.py:776]: yes, image_padding_len cannnot go negative here  because image_seq_len_target is lower-bounded by _ceil_to_multiple(image_ori_len, SEQ_MULTI_OF), which is always >= image_ori_len.

**Review by BBuf — APPROVED** [review]
